### ISSUE TITLE
Segfault when assign values to sparse matrix by multi-threads

### ISSUE THREAD

**Issue by YinLiLin** [RcppCore/RcppArmadillo#274]
Hi, 

I intend to fill a sparse matrix with values by multi-threads, I find it works fine when using 1 thread, but caught segfault for multi-threads, please see the sample codes as following, is it not applicable to do multi-threads assignment for sparse matrix? 
Looking for your suggestions. 
Thanks in advance.
```r
#include <RcppArmadillo.h>
#include <omp.h>

// [[Rcpp::plugins(cpp11)]]
// [[Rcpp::depends(RcppArmadillo)]]
// [[Rcpp::depends(bigmemory, BH)]]

using namespace std;
using namespace Rcpp;
using namespace arma;

// [[Rcpp::export]]
arma::sp_mat test(arma::vec x, int n, int threads = 1){
    omp_set_num_threads(threads);
    arma::sp_mat m(n, n);
    #pragma omp parallel for schedule(dynamic) 
    for(int i = 0; i < n; i++){
        for(int j = 0; j < n; j++){
            m(i, j) = x[i * n + j];
        }
    }
    return m;
}

# ok for 1 thread
a<-test(sample(c(0,1,2),100*100,rep=T), n=100, threads=1)
# fault for multi-threads
a<-test(sample(c(0,1,2),100*100,rep=T), n=100, threads=10)
```

**Comment by dselivanov** [comment]
Writes to sparse matrices are not thread safe.

**Comment by eddelbuettel** [comment]
I saw something similar come up for Eigen the other day (on StackOverflow).  So I think you just cannot do this.

**Comment by eddelbuettel** [comment]
Closing this [as cross-posting to SO](https://stackoverflow.com/questions/58445033/segfault-when-assign-values-to-sparse-matrix-by-multi-threads) is abusive.   

Maybe post the 2nd site after a few days if nobody has responded.

**Comment by conradsnicta** [comment]
@eddelbuettel @dselivanov - This should work with Armadillo versions 9.600 and later, as the element accessors have explicit protections against multi-threaded writes to sparse matrices:  
https://gitlab.com/conradsnicta/armadillo-code/blob/9.800.x/include/armadillo_bits/MapMat_meat.hpp#L1200-1219

@YinLiLin - I suggest you update your installation of RcppArmadillo to the current version.

**Comment by YinLiLin** [comment]
Many thanks to all of you for the comments, as @conradsnicta suggested, the problem successfully solved after updating RcppArmadill to 9.800.
Thanks again.

**Comment by eddelbuettel** [comment]
@conradsnicta excellent, thanks -- might it be worthwhile to mention this on http://arma.sourceforge.net/docs.html as well?
